src/modify.js

```diff
@@ -66,12 +67,39 @@ function rewriteAllFields(schema, configCallback, context) {
   });
 }
 
+function createFields(schema, fieldsConfig) {
+  if (!fieldsConfig) return null;
+
+  const fieldsToAdd = Object.entries(fieldsConfig);
+
+  fieldsToAdd.forEach(([shortPath, fieldAttrs]) => {
+    const fieldPath = shortToFullPath(shortPath);
+
+    if (fieldAttrs.properties) {
+      // Recursive to nested fields...
+      createFields(get(schema.properties, fieldPath), fieldAttrs.properties);
+    }
+
+    const fieldInSchema = get(schema.properties, fieldPath);
+    if (fieldInSchema) {
+      // Do not override/edit a field that already exists.
+      // That's the job of config.fields method.
```

Comment:
Similar to the pick fields, should we warn when a field creation is not done?

Something like:
```
  pickConfig.onWarn({
    message: `You've tried to create a field at ${fieldPath}, however the field path already exists in the schema. This field will be ignored.'
  });
```

Reply:
Ah good point! But now the question is the API... Where exactly do we put `onWarn`?

In the pick, the API is:

```js
modify(schema, {
  pick: {
    fields: {
        [fieldsName]: "...."
    },
    onWarn: () => "...."
  }
})
```

But the API for create was different. It accepts the fields directly. Where would we put the `onWarn()`?

```js
modify(schema, {
  create: {
    [fieldsName]: "...."
  }
})
```

Or, we could have a global onWarn to all type of warnings. 

```js
modify(schema, {
  create: {
    [fieldsName]: "...."
  },
  pick: { ... }
  onWarn: (warnings) =>
})
```

The warnings would be an array  of all multiple warnings, from pick, create, etc etc. WDYT?

Reply:
The next question is. Shall we go back to the pick MR https://github.com/remoteoss/json-schema-form/pull/79/files and change the API to be flat like `create` and `fields`?

Before:

```js
modify(schema, {
  pick: {
    fields: {
        [fieldsName]: "...."
    },
    onWarn: () => "...."
  }
})
```

After:

```js
modify(schema, {
  pick: {
    [fieldsName]: "...."
  },
  onWarn: () => "...."
  }
})
```

Reply:
Actually, the order PR https://github.com/remoteoss/json-schema-form/pull/78 also has nested `fields`:

```
const result = modify(schema, {
  order: {
    fields: ['field_c', 'field_a', 'field_b'],
    remaining: 'end'
  },
);
```

Decisions, decisions... I'm towards adding the nested `fields` attr to every main method. (create, pick, order).

@joaoalmeida-remote

Reply:
I agree having `fields` property across all the utility methos.
As for the on warn we could follow the usal Node JS function signature and accept a onError parameter (as a 3rd parameter) and so it would look like:

```
modify(schema, {
  pick: {
    fields: {
        [fieldsName]: "...."
    },
  },
  (err) => {
    console.log(err)
  } 
})
```

One of the advantages is that we could use the onError method in all the utilities methods. Thoughts?

Reply:
@eng-almeida you meant as a third parameter to the `modify` call right? So like this?

```
modify(schema, {
  pick: {
    fields: {
        [fieldsName]: "...."
    },
  }
}, (err) => { console.log(err)} )
```

I'm not sure here, putting it inside the config object gives more flexibility to later – if necessary – change the `modify` API or introduce a new callback. You can make changes to the config without consumers having to modify calls if they don't want the new behavior, e.g. deprecated a config option.

Also not sure about `onError` since we don't seem to error ever, we just warn about potentially mis-configured schemas.

Reply:
I agree with @danielcardoso5, in other words:

```js
modify(schema, {
  pick: {
    fields: {
        [fieldsName]: "...."
    },
  },
  onWarn: => {}
}
```

We don't deal with errors at all at the moment, but if we ever did, probably it's gonna be a typical try catch. 

Or, another point different from `createHeadlessForm`, there we do like this:

```js
const { fields, error } = createHeadlessForm(schema)
```

Perhaps we should do the same for modify?

```js
const { schema, error } = modify(schema, originalSchema)
```

Reply:
Do you know if there are major pros/cons of using a callback vs direct result, besides the way you access/handle it?

As I see it, with callback we can enforce it's not forgotten (TS, throwing error, etc). With the direct result, we can't.

I'm not sure which one to pick, need to give it a thought. But probably would keep it with the callback.(the first snippet in my previous comment)

Reply:
> you meant as a third parameter to the modify call right

@danielcardoso5 yes, thanks for the correction 🙌  I agree with the flexibility argument, even though IMO it's ok to release major versions with breaking changes. My only concern with the initial proposal is with the API format by adding `onWarn` callback next to the fields. IMO It doesn't feel very intuitive.

> I'm not sure which one to pick, need to give it a thought. 

If we're looking for consistency (with createHeadlessForm) maybe `const { schema, error, warn } = modify(schema, originalSchema)` is the best option.

Reply:
> even though IMO it's ok to release major versions with breaking changes.

Yeah, I'm okay with breaking changes too, this is still in beta.

> If we're looking for consistency (with createHeadlessForm)

I don't see it as a mandatory thing, to be honest. I prioritize quality over consistency among these 2 modules

Reply:
Let me call @brennj here too. If you need a recap, let us know!

Reply:
> maybe const { schema, error, warn } = modify(schema, originalSchema) is the best option.

Agree! Another reason I favor this approach over the callback pattern is that you only call `modify` once and you know the warnings/errors as soon as you make this call. The `onWarn` kind of callback pattern suggests you could potentially receive more warnings after the `modify` call and you could "listen" to them, but that's not the case.

Reply:
> As I see it, with callback we can enforce it's not forgotten (TS, throwing error, etc). With the direct result, we can't.

I think this is the main drawback of using the return approach. We can always issue a warning whenever someone calls `modify` (and give the ability to suppress it) to inform that calls to `modify` can result in invalid schemas and that we provide a `warnings` array that they can use to be informed of things that were potentially overlooked. 

Since we don't have TS right now we would need to either throw an error or warn that the `onWarn` callback was not provided.

Reply:
> The `onWarn` callback suggests you could potentially receive more warnings

100% true, this is not an async operation, so one less reason to use it!

> Since we don't have TS right now 

We do have it, kind of... we have the Type declarations file. Either way, let's go then with the returned value approach then. (this will be a breaking change). I'll make sure to document `warn` very well in the docs.

> We can always issue a warning whenever someone calls modify (and give the ability to suppress it) 

What do you mean here with "suppress"?

Reply:
> What do you mean here with "suppress"?

So what I was suggesting was to log a warning whenever there's a call to `modify`, something along the lines of:

`When using `modify` make sure you log the returned `warnings` as they provide useful debugging information…`

If the developers are already using the `warnings` array, there's no point in keeping logging that warning all the time, so they could potentially disable it by passing a flag to `modify`, example:

```
modify(schema, {
  pick: ...,
  disableLogging: true 
})
```

But now this made me realize that we can just log any warnings internally within `modify` by default, without having devs to either use a callback or the `warnings` return – although we can still provide the latter. If they really want to disable the built-in logs in local development, they could pass `disableLogging: true`. Most bundling tools already strip any logging in production builds.

Reply:
Oh I see... well, I think that will be super noisy but for a good cause.💡 I'll create an issue to tackle later, and focus now on finishing the current PRs.

I'll let you know when it's ready for a final review!

Reply:
Just chiming in from the tag, trying to get proper understanding of the thread, I believe I like what you're all proposing with returning `warn` etc from `modify`.